File: NovaVoice/python_service/nova_sonic_service.py
#!/usr/bin/env python3
"""
Nova Sonic Bidirectional Streaming Service
FastAPI service that provides WebSocket interface for Nova Sonic bidirectional streaming
"""

import asyncio
import base64
import json
import logging
import os
from typing import Dict, Any
import uuid

import boto3
from fastapi import FastAPI, WebSocket, WebSocketDisconnect
from fastapi.middleware.cors import CORSMiddleware

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

app = FastAPI(title="Nova Sonic Streaming Service", version="1.0.0")

# Enable CORS for Rails app
app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],  # In production, specify your Rails domain
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

class NovaSonicSession:
    """Manages a single Nova Sonic bidirectional streaming session"""
    
    def __init__(self, session_id: str):
        self.session_id = session_id
        self.bedrock_client = None
        self.stream = None
        self.websocket = None
        self.is_active = False
        
        # Initialize Bedrock client
        self._init_bedrock_client()
        
    def _init_bedrock_client(self):
        """Initialize AWS Bedrock Runtime client with role assumption"""
        try:
            # Try to assume the Nova Sonic role
            role_arn = os.getenv('NOVA_SONIC_ROLE_ARN', 'arn:aws:iam::302296110959:role/NovaVoice-NovaSonic-Role')
            region = os.getenv('AWS_REGION', 'us-east-1')
            
            sts_client = boto3.client('sts', region_name=region)
            assumed_role = sts_client.assume_role(
                RoleArn=role_arn,
                RoleSessionName=f"NovaVoice-Python-{self.session_id}"
            )
            
            credentials = assumed_role['Credentials']
            self.bedrock_client = boto3.client(
                'bedrock-runtime',
                region_name=region,
                aws_access_key_id=credentials['AccessKeyId'],
                aws_secret_access_key=credentials['SecretAccessKey'],
                aws_session_token=credentials['SessionToken']
            )
            
            logger.info(f"✅ Bedrock client initialized with assumed role for session {self.session_id}")
            
        except Exception as e:
            logger.error(f"❌ Failed to initialize Bedrock client: {e}")
            # Fallback to default credentials
            self.bedrock_client = boto3.client('bedrock-runtime', region_name=region)
            logger.info(f"🔄 Using default credentials for session {self.session_id}")
    
    async def start_stream(self, websocket: WebSocket):
        """Start bidirectional stream with Nova Sonic"""
        self.websocket = websocket
        
        try:
            logger.info(f"🚀 Starting Nova Sonic bidirectional stream for session {self.session_id}")
            
            # Note: The Python SDK also doesn't have invoke_model_with_bidirectional_stream yet
            # This is a placeholder for when it becomes available
            # For now, we'll simulate the bidirectional streaming behavior
            
            # Send session start confirmation
            await self.websocket.send_text(json.dumps({
                "type": "session_started",
                "session_id": self.session_id,
                "message": "Nova Sonic session started (simulated)"
            }))
            
            self.is_active = True
            logger.info(f"✅ Session {self.session_id} started successfully")
            
        except Exception as e:
            logger.error(f"❌ Failed to start session {self.session_id}: {e}")
            await self.websocket.send_text(json.dumps({
                "type": "error",
                "message": f"Failed to start session: {str(e)}"
            }))
    
    async def process_audio(self, audio_data: str):
        """Process incoming audio data and send to Nova Sonic"""
        if not self.is_active:
            logger.warning(f"⚠️ Session {self.session_id} not active, ignoring audio")
            return
        
        try:
            # Decode base64 audio
            audio_bytes = base64.b64decode(audio_data)
            logger.info(f"🎵 Processing audio chunk: {len(audio_bytes)} bytes for session {self.session_id}")
            
            # TODO: When bidirectional streaming is available, send audio to Nova Sonic here
            # For now, simulate Esther's response
            await self._simulate_esther_response()
            
        except Exception as e:
            logger.error(f"❌ Error processing audio for session {self.session_id}: {e}")
            await self.websocket.send_text(json.dumps({
                "type": "error",
                "message": f"Audio processing failed: {str(e)}"
            }))
    
    async def _simulate_esther_response(self):
        """Simulate Esther's scheduling-focused response"""
        import random
        
        responses = [
            "Hello! I'd love to schedule a 15-minute meeting with Mike Lawrence about our Gospel outreach program. Would this week or next work?",
            "I understand. Mike Lawrence can explain that better in our 15-minute meeting. What day works for your pastor?",
            "Perfect! Mike Lawrence will show you how this reaches unchurched families. Would mornings or afternoons work better?",
            "The meeting is between your Pastor and Mike Lawrence, our founder. I'm just scheduling it. What's your preferred time?",
            "That's exactly what Mike Lawrence can address in 15 minutes. Would you prefer Zoom or another platform?",
            "Great question! Mike Lawrence has the details on our track record. Could we schedule 15 minutes this week?",
            "I'd be happy to follow up, but Mike Lawrence can demonstrate the impact much better. What's your pastor's preferred time?",
            "For more information, visit globaloutreachevent.com or call Mike Lawrence directly at 347-300-5533. When works best for a meeting?"
        ]
        
        response_text = random.choice(responses)
        
        # Generate simple audio tone as placeholder
        audio_data = self._generate_simple_audio_beep()
        
        # Send response back to Rails via WebSocket
        await self.websocket.send_text(json.dumps({
            "type": "audio_response",
            "audio": base64.b64encode(audio_data).decode('utf-8'),
            "text": response_text,
            "session_id": self.session_id
        }))
        
        logger.info(f"💬 Sent simulated response for session {self.session_id}: {response_text[:50]}...")
    
    def _generate_simple_audio_beep(self) -> bytes:
        """Generate a simple audio beep (1-second 440Hz tone)"""
        import struct
        import math
        
        sample_rate = 16000
        duration = 1.0
        frequency = 440.0
        
        audio_data = b''
        for i in range(int(sample_rate * duration)):
            # Generate sine wave
            sample = int(math.sin(2 * math.pi * frequency * i / sample_rate) * 0.3 * 32767)
            # Convert to 16-bit PCM bytes (little endian)
            audio_data += struct.pack('<h', sample)
        
        return audio_data
    
    async def close(self):
        """Close the streaming session"""
        self.is_active = False
        if self.stream:
            # TODO: Close actual stream when available
            pass
        logger.info(f"🧹 Session {self.session_id} closed")

# Global session manager
sessions: Dict[str, NovaSonicSession] = {}

@app.websocket("/ws/nova-sonic")
async def nova_sonic_websocket(websocket: WebSocket):
    """WebSocket endpoint for Nova Sonic bidirectional streaming"""
    session_id = str(uuid.uuid4())
    session = NovaSonicSession(session_id)
    sessions[session_id] = session
    
    await websocket.accept()
    logger.info(f"🔌 WebSocket connected for session {session_id}")
    
    try:
        # Start the Nova Sonic stream
        await session.start_stream(websocket)
        
        # Listen for incoming messages
        while True:
            message = await websocket.receive_text()
            data = json.loads(message)
            
            if data.get("type") == "audio":
                await session.process_audio(data.get("audio", ""))
            elif data.get("type") == "close":
                break
            else:
                logger.warning(f"⚠️ Unknown message type: {data.get('type')}")
                
    except WebSocketDisconnect:
        logger.info(f"🔌 WebSocket disconnected for session {session_id}")
    except Exception as e:
        logger.error(f"❌ WebSocket error for session {session_id}: {e}")
    finally:
        # Cleanup
        await session.close()
        if session_id in sessions:
            del sessions[session_id]
        logger.info(f"🧹 Session {session_id} cleaned up")

def _generate_simple_audio_beep() -> bytes:
    """Generate a simple audio beep (1-second 440Hz tone)"""
    import struct
    import math
    
    sample_rate = 16000
    duration = 1.0
    frequency = 440.0
    
    audio_data = b''
    for i in range(int(sample_rate * duration)):
        # Generate sine wave
        sample = int(math.sin(2 * math.pi * frequency * i / sample_rate) * 0.3 * 32767)
        # Convert to 16-bit PCM bytes (little endian)
        audio_data += struct.pack('<h', sample)
    
    return audio_data

@app.get("/health")
async def health_check():
    """Health check endpoint"""
    return {"status": "healthy", "active_sessions": len(sessions)}

@app.get("/")
async def root():
    """Root endpoint"""
    return {"message": "Nova Sonic Streaming Service", "version": "1.0.0"}

if __name__ == "__main__":
    import uvicorn
    
    port = int(os.getenv("PORT", 8001))
    host = os.getenv("HOST", "0.0.0.0")
    
    logger.info(f"🚀 Starting Nova Sonic service on {host}:{port}")
    uvicorn.run(app, host=host, port=port, log_level="info")

File: NovaVoice/app/services/nova_sonic_service.rb
# app/services/nova_sonic_service.rb - Real Nova Sonic Implementation
require 'json'
require 'base64'
require 'securerandom'
require 'aws-sdk-bedrockruntime'
require 'aws-sdk-sts'

class NovaSonicService
  MODEL_ID = 'amazon.nova-sonic-v1:0'
  REGION = ENV['AWS_DEFAULT_REGION'] || ENV['AWS_REGION'] || 'us-east-1'

  def initialize(channel)
    @channel = channel
    @session_id = SecureRandom.uuid
    @processing = false
    @mutex = Mutex.new
    @bedrock_client = nil
    
    Rails.logger.info "🎤 NovaSonicService initialized with session: #{@session_id}"
    Rails.logger.info "🌍 Using AWS region: #{REGION}"
    
    # Initialize AWS Bedrock Runtime client
    initialize_bedrock_client
  end

  def process_audio(audio_data)
    Rails.logger.info "🎵 Processing audio chunk: #{audio_data.length} bytes"
    
    @mutex.synchronize do
      if @bedrock_client
        # Process with REAL Nova Sonic
        process_with_nova_sonic(audio_data)
      else
        Rails.logger.error "❌ Bedrock client not available"
        @channel.broadcast_error("Nova Sonic service unavailable")
      end
    end
  end

  def cleanup
    Rails.logger.info "🧹 Cleaning up Nova Sonic service"
    @mutex.synchronize do
      @processing = false
    end
  end

  private

  def esther_system_prompt
    <<~PROMPT
      # Esther - Scheduling Assistant for Mike Lawrence Productions

      ## Role
      You are Esther, Mike Lawrence Productions' scheduling assistant. Your ONLY job is to schedule 15-minute web meetings between senior pastors and Mike Lawrence about our Gospel outreach program.

      ## Key Facts
      - **Program**: Two-phase outreach (entertainment THEN Gospel presentation)
      - **Format**: 40-50 min Off-Broadway illusion show + 30 min separate Gospel message
      - **Track Record**: Similar to Campus Crusade approach (~100,000 decisions)
      - **Your Role**: Schedule meetings ONLY - you do NOT attend meetings

      ## Meeting Arrangement
      **CRITICAL**: You schedule meetings between the pastor and Mike Lawrence. You are NOT a meeting participant.

      **When asked who attends**: "The meeting is between your Pastor and Mike Lawrence, our founder. I'm just scheduling it for you."

      ## Conversation Style
      - **Be Brief**: 1-2 sentences maximum per response
      - **Stay Focused**: Always redirect to scheduling the meeting
      - **Be Professional**: Warm but concise

      ## Critical Information
      **Website**: globaloutreachevent.com
      **Mike Lawrence Direct Number**: 347-300-5533
      **Meeting Participants**: Pastor + Mike Lawrence (NOT Esther)

      ## Response Pattern
      1. **Acknowledge briefly** (1-2 words)
      2. **Redirect to Mike Lawrence meeting** (1 sentence)
      3. **Ask for scheduling** (1 question)

      Keep all responses under 25 words when possible. Always redirect to scheduling the meeting with Mike Lawrence.
    PROMPT
  end


  def initialize_bedrock_client
    begin
      role_arn = ENV['NOVA_SONIC_ROLE_ARN'] || 'arn:aws:iam::302296110959:role/NovaVoice-NovaSonic-Role'
      
      # Try to assume the Nova Sonic role first
      begin
        sts_client = Aws::STS::Client.new(region: REGION)
        assume_role_response = sts_client.assume_role(
          role_arn: role_arn,
          role_session_name: "NovaVoice-#{@session_id}"
        )
        
        credentials = Aws::Credentials.new(
          assume_role_response.credentials.access_key_id,
          assume_role_response.credentials.secret_access_key,
          assume_role_response.credentials.session_token
        )
        
        Rails.logger.info "✅ Successfully assumed Nova Sonic role: #{role_arn}"
        Rails.logger.info "🔐 Access Key: #{credentials.access_key_id[0..8]}..."
        
        # Create Bedrock Runtime client with assumed role credentials
        @bedrock_client = Aws::BedrockRuntime::Client.new(
          region: REGION,
          credentials: credentials
        )
        
      rescue => role_error
        Rails.logger.warn "⚠️ Failed to assume role #{role_arn}: #{role_error.message}"
        
        # Fallback to default credential chain
        @bedrock_client = Aws::BedrockRuntime::Client.new(region: REGION)
        Rails.logger.info "🔄 Using fallback credentials with default client"
      end
      
      Rails.logger.info "✅ Bedrock Runtime client initialized successfully"
      
    rescue => e
      Rails.logger.error "❌ Failed to initialize Bedrock client: #{e.message}"
      @bedrock_client = nil
    end
  end

  def process_with_nova_sonic(audio_data)
    Thread.new do
      begin
        Rails.logger.info "🚀 Calling REAL Nova Sonic with invoke_model_with_response_stream"
        
        # Convert WebM to PCM if needed
        pcm_audio = convert_to_pcm(audio_data)
        
        # Prepare the request body for Nova Sonic
        request_body = {
          inputText: "",  # Nova Sonic can work with just audio
          inputAudio: Base64.encode64(pcm_audio),
          audioConfig: {
            format: "pcm",
            sampleRateHertz: 16000
          },
          inferenceConfig: {
            maxTokens: 1024,
            temperature: 0.7,
            topP: 0.9
          }
        }
        
        # Add system prompt
        if esther_system_prompt && !esther_system_prompt.empty?
          request_body[:inputText] = esther_system_prompt
        end
        
        Rails.logger.info "📡 Sending request to Nova Sonic: #{MODEL_ID}"
        
        # Call Nova Sonic with response streaming
        response = @bedrock_client.invoke_model_with_response_stream(
          model_id: MODEL_ID,
          body: JSON.generate(request_body),
          content_type: 'application/json'
        )
        
        Rails.logger.info "✅ Nova Sonic response stream received"
        
        # Process the streaming response
        process_nova_sonic_response_stream(response)
        
      rescue => e
        Rails.logger.error "❌ Error calling Nova Sonic: #{e.class} - #{e.message}"
        Rails.logger.error e.backtrace.first(5).join("\n")
        @channel.broadcast_error("Nova Sonic error: #{e.message}")
      end
    end
  end

  def process_nova_sonic_response_stream(response)
    audio_chunks = []
    text_response = ""
    
    response.body.each do |event|
      if event.respond_to?(:chunk) && event.chunk
        begin
          chunk_data = JSON.parse(event.chunk.bytes.read)
          Rails.logger.debug "📦 Nova Sonic chunk: #{chunk_data.keys.join(', ')}"
          
          # Handle different response types
          if chunk_data['outputAudio']
            audio_data = Base64.decode64(chunk_data['outputAudio'])
            audio_chunks << audio_data
            Rails.logger.info "🔊 Received audio chunk: #{audio_data.length} bytes"
          end
          
          if chunk_data['outputText']
            text_response += chunk_data['outputText']
            Rails.logger.info "💬 Received text: #{chunk_data['outputText']}"
          end
          
        rescue JSON::ParserError => e
          Rails.logger.error "❌ Failed to parse Nova Sonic chunk: #{e.message}"
        end
      end
    end
    
    # Combine all audio chunks and send response
    if audio_chunks.any?
      combined_audio = audio_chunks.join
      @channel.broadcast_audio(combined_audio, text_response)
      Rails.logger.info "✅ Sent Nova Sonic response: #{combined_audio.length} bytes audio, #{text_response.length} chars text"
    else
      Rails.logger.warn "⚠️ No audio received from Nova Sonic, sending text only"
      @channel.broadcast_audio(nil, text_response.empty? ? "I'm here to help schedule your meeting with Mike Lawrence." : text_response)
    end
  end
  
  def simulate_esther_response
    # Add a small delay to simulate thinking
    Thread.new do
      sleep(1.5)
      
      # Simulate Esther's scheduling-focused response
      responses = [
        "Hello! I'd love to schedule a 15-minute meeting with Mike Lawrence about our Gospel outreach program. Would this week or next work?",
        "I understand. Mike Lawrence can explain that better in our 15-minute meeting. What day works for your pastor?",
        "Perfect! Mike Lawrence will show you how this reaches unchurched families. Would mornings or afternoons work better?",
        "The meeting is between your Pastor and Mike Lawrence, our founder. I'm just scheduling it. What's your preferred time?",
        "That's exactly what Mike Lawrence can address in 15 minutes. Would you prefer Zoom or another platform?",
        "Great question! Mike Lawrence has the details on our track record. Could we schedule 15 minutes this week?",
        "I'd be happy to follow up, but Mike Lawrence can demonstrate the impact much better. What's your pastor's preferred time?",
        "For more information, visit globaloutreachevent.com or call Mike Lawrence directly at 347-300-5533. When works best for a meeting?"
      ]
      
      response_text = responses.sample
      
      # Generate simple audio beep as placeholder for voice
      audio_data = generate_simple_audio_beep
      
      # Broadcast both audio and text response
      @channel.broadcast_audio(audio_data, response_text)
      
      Rails.logger.info "💬 Simulated Esther response: #{response_text[0..50]}..."
    end
  end
  
  def generate_simple_audio_beep
    # Generate a simple audio beep to indicate voice response
    # This creates a 1-second 440Hz tone (A4 note)
    sample_rate = 16000
    duration = 1.0
    frequency = 440.0
    
    audio_data = []
    (sample_rate * duration).to_i.times do |i|
      # Generate sine wave
      sample = (Math.sin(2 * Math::PI * frequency * i / sample_rate) * 0.3 * 32767).to_i
      # Convert to 16-bit PCM bytes (little endian)
      audio_data << [sample].pack('s<')
    end
    
    audio_data.join
  end

  
  def convert_to_pcm(webm_audio_data)
    AudioConverterService.webm_to_pcm(webm_audio_data)
  end
  
  def process_response_chunk(chunk)
    begin
      event_data = JSON.parse(chunk)
      handle_nova_sonic_event(event_data)
    rescue JSON::ParserError => e
      Rails.logger.error "❌ Failed to parse response chunk: #{e.message}"
      Rails.logger.debug "🔍 Raw chunk: #{chunk[0..100]}..." if chunk.length > 100
    rescue => e
      Rails.logger.error "❌ Error processing response chunk: #{e.message}"
    end
  end
  
  def process_bidirectional_responses(stream)
    Thread.new do
      begin
        Rails.logger.info "👂 Processing Nova Sonic bidirectional responses"
        
        stream.each do |chunk|
          begin
            event_data = JSON.parse(chunk)
            handle_nova_sonic_event(event_data)
          rescue JSON::ParserError => e
            Rails.logger.error "❌ Failed to parse response chunk: #{e.message}"
          end
        end
        
      rescue StandardError => e
        Rails.logger.error "❌ Error processing bidirectional responses: #{e.message}"
        # Fallback to simulation
        simulate_esther_response
      end
    end
  end
  

end